woocommerce/google-listings-and-ads
# Product Block Editor: Add a custom block for editing channel visibility

### Changes proposed in this Pull Request:

To be compatible with Product Block Editor, this PR adds a custom block for editing channel visibility.

### Screenshots:

#### 📷 Sync status is `synced` or `null`

![1](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/747dd9ff-48b7-447d-aab4-7f833ab0c752)

#### 📷 Sync status is `pending`

![2](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/b2a5ccd1-9651-4405-be40-0c5d20a31751)

#### 📷 Sync status is `not-synced`

![3](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/2222aa6f-b7b6-4714-a9f8-89f13bfe1075)

#### 📷 Sync status is `has-errors`

![4](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/f87ade64-375e-40dd-a637-b7aeb6b374a8)

#### 📷 Variation product should not show the channel visibility

![5](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/db360523-4e55-4422-8f32-ad60de503e78)

#### 📹 Channel visibility is disabled when the product is hidden from the product catalog

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/38fe3f30-12ff-4329-b23e-fb9ef2dfe852

### Detailed test instructions:

#### 📌 Prepare test environment

1. Please refer to [the test preparation in PR 2151](https://github.com/woocommerce/google-listings-and-ads/pull/2151#prepare-test-environment) with **Woo 8.5.1**.
1. `npm install`
1. `npm start`

#### 📌 Test with simple product and disabled state

1. Go to edit a simple product, e.g. the **Beanie with Logo** imported from WC's sample products.
1. Switch to the **Google Listings & Ads** tab.
1. Change the selected channel visibility option and click the **Update** button at the top-right of the page.
1. Switch to the **Organization** tab.
1. Tick the "Hide in product catalog" option in the **Product catalog** section.
1. Click the **Update** button.
1. Switch to the **Google Listings & Ads** tab.
1. See if the channel visibility selection is disabled.
1. Switch to the **Organization** tab.
1. Uncheck the "Hide in product catalog" option.
1. Switch to the **Google Listings & Ads** tab.
1. Click the **Update** button.
1. See if the channel visibility selection is changed from disabled to enabled after updating.

💡 It must update the product data first because the `is_visible` value in the wp-data store won't be recalculated after changing the product's `catalog_visibility ` (the "Hide in product catalog" option) in the wp-data store. Instead, the `is_visible` value is only updated after the product update issues API request and syncs the response to the wp-data store.

#### 📌 Test with variable and variation products

1. Go to edit a variable product, e.g. the **Hoodie** imported from WC's sample products.
1. Switch to the **Google Listings & Ads** tab.
1. Change the selected channel visibility option and click the **Update** button.
1. Switch to the **Variations** tab.
1. Click one of the **Edit** links within the **Variations** section to edit a variation product.
1. Check if the channel visibility section is not shown.

#### 📌 Test the `has-errors` status

1. Go to edit a simple product.
1. Switch to the **Pricing** tab.
1. Clear the list price and sale price.
1. Switch to the **Google Listings & Ads** tab.
1. Enter an invalid GTIN such as `aaaa`.
1. Select "Sync and show" for the channel visibility.
1. Click the **Update** button.
1. Open the Connection Test: `/wp-admin/admin.php?page=connection-test-admin-page`.
1. Use the **Sync Product** to sync the above product
   - You might need to make the [MerchantCenterService::is_ready_for_syncing](https://github.com/woocommerce/google-listings-and-ads/blob/20169c47a3cfa2ef9c588ea69684688d4195c677/src/MerchantCenter/MerchantCenterService.php#L106) always return `true` if using a local test site.
1. Back to the product editor and refresh the webpage.
1. Check if it shows **Issues**.
   ![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/4ef5ef1d-880b-4d31-b5e2-88518fd3e12e)

### Changelog entry
